### PR TITLE
Infrastructure/refactor ledger updates

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/application/EpochsNodeSelector.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+public class EpochsNodeSelector implements NodeSelector {
+	private final Random random = new Random();
+
+	@Override
+	public Single<BFTNode> nextNode(RunningNetwork network) {
+		return network.latestEpochChanges()
+			.map(e -> {
+				ImmutableList<BFTValidator> validators = e.getBFTConfiguration()
+					.getValidatorSet().getValidators().asList();
+				int validatorSetSize = validators.size();
```
Comment (siy) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/application/EpochsNodeSelector.java:39]: Long call chain suggests that we're looking too deep into class internals. Probably we should have method size() in BFTValidatorSet. There are several locations where this call sequence is repeated, so we could reduce boilerplate and improve readability.

**Review by siy — CHANGES_REQUESTED** [review]

File: radixdlt/src/main/java/com/radixdlt/consensus/epoch/EpochManager.java [hunk 2]
```diff
@@ -112,35 +113,44 @@
 		void sendTimeoutProcessed(Timeout timeout);
 	}
 
+	private final Logger logger = LogManager.getLogger();
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/epoch/EpochManager.java]: Why is this not static?  No need to create for each instance.

File: radixdlt/src/main/java/com/radixdlt/consensus/sync/VertexStoreBFTSyncRequestProcessor.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Processor of sync requests and responds with info from a VertexStore
+ */
+public final class VertexStoreBFTSyncRequestProcessor implements BFTSyncRequestProcessor {
+	private final Logger log = LogManager.getLogger();
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/sync/VertexStoreBFTSyncRequestProcessor.java]: Why is this not static?

File: radixdlt/src/main/java/com/radixdlt/consensus/sync/VertexStoreSync.java [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+		 * will be returned.
+		 *
+		 * @param node the node to retrieve the vertex info from
+		 * @param id the id of the vertex to retrieve
+		 * @param count number of vertices to retrieve
+		 */
+		void sendGetVerticesRequest(BFTNode node, Hash id, int count);
+	}
+
+	private final Logger log = LogManager.getLogger();
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/sync/VertexStoreSync.java]: Why is this not static?

File: radixdlt/src/main/java/com/radixdlt/consensus/epoch/EpochManager.java [hunk 5]
```diff
@@ -220,18 +241,25 @@ private long currentEpoch() {
 		return this.currentEpoch.getEpoch();
 	}
 
-	public void processEpochChange(EpochChange epochChange) {
+	public void processLedgerUpdate(EpochsLedgerUpdate epochsLedgerUpdate) {
+		epochsLedgerUpdate.getEpochChange().ifPresentOrElse(
+			this::processEpochChange,
+			() -> this.syncLedgerUpdateProcessor.processLedgerUpdate(epochsLedgerUpdate)
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/epoch/EpochManager.java:247]: Ha!  Noice.

File: radixdlt/src/main/java/com/radixdlt/consensus/epoch/EpochManager.java (hunk 2 above)
Comment (clstrfsck) [reply radixdlt/src/main/java/com/radixdlt/consensus/epoch/EpochManager.java]: Also not sure why renamed from log to logger?  This is cluttering up the diff.

File: radixdlt/src/main/java/com/radixdlt/consensus/sync/GetVerticesRequest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+package com.radixdlt.consensus.sync;
+
+import com.radixdlt.consensus.bft.BFTNode;
+import com.radixdlt.crypto.Hash;
+import java.util.Objects;
+
+/**
+ * A request for bft vertices info
+ */
+public final class GetVerticesRequest {
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/sync/GetVerticesRequest.java:27]: Pojo should probably have hashCode, equals and toString.  But definitely toString.

File: radixdlt/src/main/java/com/radixdlt/consensus/sync/VertexStoreSync.java [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+ */
+public final class VertexStoreSync implements BFTSyncResponseProcessor, BFTUpdateProcessor, BFTSyncer, LedgerUpdateProcessor<LedgerUpdate> {
+	private enum SyncStage {
+		PREPARING,
+		GET_COMMITTED_VERTICES,
+		SYNC_TO_COMMIT,
+		GET_QC_VERTICES
+	}
+
+	private static class SyncState {
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/sync/VertexStoreSync.java:63]: Will probably need to refactor this for TCs.

File: radixdlt/src/test/java/com/radixdlt/EpochsConsensusModuleTest.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+import com.radixdlt.counters.SystemCounters;
+import com.radixdlt.network.TimeSupplier;
+import org.junit.Test;
+
+public class EpochsConsensusModuleTest {
+	@Inject
+	private BFTInfoSender infoSender;
+
+	@Inject
+	private TimeoutSender timeoutSender;
```
Comment (clstrfsck) [inline radixdlt/src/test/java/com/radixdlt/EpochsConsensusModuleTest.java]: Not used?

File: radixdlt/src/test/java/com/radixdlt/ConsensusModuleTest.java [hunk 9]
```diff
@@ -17,57 +17,104 @@
...
-			bind(NextCommandGenerator.class).toInstance(mock(NextCommandGenerator.class));
-			bind(SystemCounters.class).toInstance(mock(SystemCounters.class));
-			bind(TimeSupplier.class).toInstance(mock(TimeSupplier.class));
-			bind(Hasher.class).toInstance(mock(Hasher.class));
-			bind(HashVerifier.class).toInstance(mock(HashVerifier.class));
-			bind(HashSigner.class).toInstance(mock(HashSigner.class));
-			bind(BFTNode.class).annotatedWith(Names.named("self")).toInstance(mock(BFTNode.class));
-		}
+	@Inject
+	private BFTFactory bftFactory;
```
Comment (clstrfsck) [inline radixdlt/src/test/java/com/radixdlt/ConsensusModuleTest.java]: Not used?

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicTest.java [hunk 10]
```diff
@@ -80,13 +76,23 @@ private DeterministicTest(
 	}
 
 	public static class Builder {
+		private enum LedgerType {
+			MOCKED_LEDGER(false),
+			LEDGER_AND_EPOCHS_AND_SYNC(true);
+
+			private final boolean hasEpochs;
```
Comment (clstrfsck) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicTest.java]: Could remove this for now, as it is not used.

**Review by clstrfsck — CHANGES_REQUESTED** [review]
Looks OK, some minor nits, but I think the non-static logger thing needs to be fixed.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/application/EpochsNodeSelector.java (hunk 1 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/application/EpochsNodeSelector.java:39]: Agreed though I'd rather do this change in another PR

**Review by clstrfsck — APPROVED** [review]

**Pull request merged**